test/continuous/test_bcq.py

```diff
@@ -0,0 +1,341 @@
+import argparse
+import datetime
+import os
+import pprint
+
+import gym
+import numpy as np
+import torch
+from torch.utils.tensorboard import SummaryWriter
+
+from tianshou.data import Collector, VectorReplayBuffer
+from tianshou.env import DummyVectorEnv, SubprocVectorEnv
+from tianshou.policy import BCQPolicy, SACPolicy
+from tianshou.policy.imitation.bcq import VAE, Perturbation
+from tianshou.trainer import offpolicy_trainer
+from tianshou.utils import BasicLogger, TensorboardLogger
+from tianshou.utils.net.common import MLP, Net
+from tianshou.utils.net.continuous import ActorProb, Critic
+
+
+def get_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--task', type=str, default='Pendulum-v0')
+    parser.add_argument('--seed', type=int, default=0)
+    parser.add_argument('--buffer_size', type=int, default=20000)
+    parser.add_argument('--sac_hidden_sizes', type=int, nargs='*', default=[128, 128])
+    parser.add_argument('--hidden_sizes', type=int, nargs='*', default=[200, 150])
+    parser.add_argument('--actor_lr', type=float, default=1e-3)
+    parser.add_argument('--critic_lr', type=float, default=1e-3)
+    parser.add_argument("--start_timesteps", type=int, default=50000)
+    parser.add_argument('--epoch', type=int, default=7)
+    parser.add_argument('--step_per_epoch', type=int, default=8000)
+    parser.add_argument('--batch_size', type=int, default=256)
+    parser.add_argument('--training_num', type=int, default=10)
+    parser.add_argument('--test_num', type=int, default=10)
+    parser.add_argument('--step_per_collect', type=int, default=10)
+    parser.add_argument('--update_per_step', type=float, default=0.125)
+    parser.add_argument('--logdir', type=str, default='log')
+    parser.add_argument('--render', type=float, default=0.)
+
+    parser.add_argument("--vae_hidden_sizes", type=int, nargs='*', default=[375, 375])
+    parser.add_argument("--gamma", default=0.99)
+    parser.add_argument("--tau", default=0.005)
+    # Weighting for Clipped Double Q-learning in BCQ
+    parser.add_argument("--lmbda", default=0.75)
+    # Max perturbation hyper-parameter for BCQ
+    parser.add_argument("--phi", default=0.05)
+    parser.add_argument(
+        '--device', type=str, default='cuda' if torch.cuda.is_available() else 'cpu'
+    )
+    parser.add_argument('--resume_path', type=str, default=None)
+    parser.add_argument(
+        '--watch',
+        default=False,
+        action='store_true',
+        help='watch the play of pre-trained policy only'
+    )
+    # sac:
+    parser.add_argument('--alpha', type=float, default=0.2)
+    parser.add_argument('--auto_alpha', type=int, default=1)
+    parser.add_argument('--alpha_lr', type=float, default=3e-4)
+    parser.add_argument('--rew_norm', action="store_true", default=False)
+    parser.add_argument('--n_step', type=int, default=3)
+    args = parser.parse_known_args()[0]
+    return args
+
+
+def gather_data():
```

Comment:
This part has a lot of duplicated code. I suggest reusing the existing code (and also using some features of pytest) to finish data generation (save to .pkl / .h5 format and load from that dataset).

Reply:
for example https://github.com/RKrahl/pytest-dependency